src/core/mac/mac_header_ie.hpp

```diff
@@ -326,20 +444,15 @@ class RendezvousTimeIe
  * Implements Connection IE data structure.
  */
 OT_TOOL_PACKED_BEGIN
-class ConnectionIe : public VendorIeHeader
+class ConnectionIe : public ThreadIe::Trait<ThreadIe::SubType::kConnection>
 {
 public:
-    static constexpr uint8_t kHeaderIeId      = ThreadIe::kHeaderIeId;
-    static constexpr uint8_t kIeContentSize   = ThreadIe::kIeContentSize + sizeof(uint8_t);
-    static constexpr uint8_t kThreadIeSubtype = 0x01;
-
     /**
      * Initializes the Connection IE.
      */
-    void Init(void)
+    ConnectionIe(void)
+        : mThreadIe(kSubType)
```

Comment:
This PR changed the different `Ie` specialized classes to use a constructor for initialization (instead of an `Init()` method). While this can work, it has some downsides.

A constructor will always be used, even if we declare a `ConnectionIe connIe`, which we intend to read and populate from a received frame or message (by copying the bytes).

We still want to use the templated `Init()` in `HeaderIe`, which does the following:

```c++
template <typename ContentType> ContentType *Init(uint8_t aLen)
    {
        static_assert(std::is_standard_layout<ContentType>::value, "ContentType is not standard layout");
        Init(ContentType::kHeaderIeId, aLen);
        return new (this + 1) ContentType();
```

The use of `new (this + 1) ContentType()` works correctly, but it is more complex to follow and read. It also makes it harder for toolchains to optimize or inline this code.

Should we keep the `Init()` method instead?